### ISSUE TITLE
deadlock when _in is a blocking file

### ISSUE THREAD

**Issue by omribahumi** [amoffat/sh#253]
When running the following code:

```python
import sys
from sh import uptime

print uptime(_in=sys.stdin)
```

It blocks forever.

Will also happen when doing:
```python
import os
from sh import uptime

r, w = os.pipe()
print uptime(_in=os.fdopen(r, 'r'))
```

I don't see any way to fix this without refactoring the entire code (to support redirection of actual OS FDs), but I think this behavior should be documented, hence I'm opening a ticket.

**Comment by omribahumi** [comment]
When I come to think of it, perhaps doing a `select()` call on `_in` instead of doing a blocking `.read()` would be possible. Haven't looked at the code though.

**Comment by amoffat** [comment]
Here's the relevant code if you wanted to experiment:

https://github.com/amoffat/sh/blob/master/sh.py#L1710 `get_file_chunk_reader` returns a function that knows how to read from a fd.  The `select()` could be performed here, and if the fd wasn't ready, raise a `NotYetReadyToRead`.  In theory that should be all that needs to happen, but I'd be interested to see if it actually works and tests don't break

**Comment by amoffat** [comment]
this is fixed on the release 1.2 branch.  thanks for reporting @omribahumi
